### ISSUE TITLE
How to get a document name of to be included file?

### ISSUE THREAD

**Issue by AlexanderZobkov** [asciidoctor/asciidoctor#2608]
Hello,

There is a master file that includes (with multiple instances of a standard include directive) multiple smaller and different parts/files. 
Is it possible to get a name of the smaller part/file by referring to an attribute? Experiments show when referring to the attribute 'docfile' in smaller parts/files, it always returns a name of the master file.
If not possible by referring to an attribute, how this potentially could be implemented with extensions?

master.adoc
```
= My book

include::chapter01.adoc[]

Source document {docfile}

...

include::chapterXX.adoc[]
```

chapter01.adoc
```

== Section

Source document {docfile} <- returns the same as in main.adoc 

```

**Comment by diguage** [comment]
It is a great problem.

**Comment by elextr** [comment]
`{infile}` ?

**Comment by mojavelinux** [comment]
infile is not implemented in Asciidoctor. I will explain why and suggest another way to get this information.

**Comment by mojavelinux** [comment]
The reason infile and indir are not implemented in Asciidoctor is because there's no place to reliably insert this information into the document. An include directive is a preprocessor, which means that it can occur anywhere in the document and has no awareness of the surrounding context. In other words, there's no node to attach this information to.

The reason this worked in AsciiDoc Py is because AsciiDoc Py is a streaming processor and it converts while it parses. So the information can remain in scope as the document is processed. Asciidoctor parses to an AST, so the local scope is gone.

**Comment by mojavelinux** [comment]
You might be interested in this extension: https://rubygems.org/gems/asciidoctor-indir_ext

**Comment by AlexanderZobkov** [comment]
Hello @rdmueller, 

if you use gradle in your projects, you could add a task to inject file name.
```
processDocsResources {
    into sourceSets.docs.output.resourcesDir
    eachFile {
        it.filter(ReplaceTokens, tokens: [fileName: it.name], beginToken: '@{', endToken: '}')
    }
}
```

**Comment by mojavelinux** [comment]
I don't think the question is about the main file. That's already available. It's whether the name of the file being included somewhere in the middle of the document is available to content inside that file.

**Comment by AlexanderZobkov** [comment]
The provided snippet injects into a file (be it main one or some part) its own name.  See eachFile and it.name.

**Comment by mojavelinux** [comment]
Ah, I see. :+1:

**Comment by rdmueller** [comment]
@docToolchain also uses Gradle to prepend the filename to each `.adoc` file:
https://github.com/docToolchain/docToolchain/blob/master/scripts/prependFilename.gradle

this is needed to implement the "improve this doc" and "create an issue" buttons for example on https://doctoolchain.github.io/docToolchain/

These buttons link directly to the git repository - the "improve this doc" directly to the `.adoc` file and the "create an issue" button uses the filename to automatically include it as reference to the issue.

**Comment by lopsotronic** [comment]
This is a hard stop for my pipeline as well. The included filenames need to be identified in the output for purposes of revision control and feedback. One workaround - declaring filename as user defined atttribute in running content- doesn't work in asciidoctor-pdf, as it uses just the first instance of the attribute for running content (header/footer). DocBook can use //abstract from each subsection, but it's nasty.

**Comment by mojavelinux** [comment]
The include directive fuses the content with the original document, so there's no place to store the filename (as the include can be anywhere, including across the boundaries of a node). If you want to capture and inject the filename somewhere, you can implement a custom include processor.

**Comment by rdmueller** [comment]
what if we extend the standard include processor to prepends a
`:includefilename: xyz`
to each included section?
I guess it would do no harm but solve this issue.

**Comment by mojavelinux** [comment]
As I was explaining, since the include directive can be used anywhere, there's no guarantee that it's a valid place for an attribute entry, that it will be interpreted, or that it won't break the document. (For example, it could be inside a verbatim block or between list items). In fact, the include directive doesn't see the structure of the document at all. So yes, it will do harm.

That's why I think the extension is the right solution here. If you have control over the include and how it is used and processed, then you can safely add such an attribute entry.

**Comment by rdmueller** [comment]
thanx for the explanation. now I do understand and also understand why it works for me (I only prepend for `.adoc` files I include and these are always full docs and not parts.

Does it make sense to already collect requirements like this for the spec to be created by the workinggroup?

**Comment by mojavelinux** [comment]
Indeed. Though what we're talking about here is covered by the subdocument macro, which is already flagged for spec. See #3603.

The include directive should remain void of context because, by it's very nature, it's a preprocessor directive and is supposed to be context free.

**Comment by rdmueller** [comment]
so, if we design it properly, shouldn't the `{filename}` attribute be even independant of any macro?

**Comment by mojavelinux** [comment]
The issue is where to define it. A subdocument macro can define it because it has access to the document model. A preprocessor directive is merging raw lines together. It knows nothing about the document tree (as it hasn't been parsed yet) and has no where to put that information to make it available to the document itself. It's just not the right tool for this (unless you can enforce constraints, as an extension could do).

**Comment by rdmueller** [comment]
ok. so, a new macro seems to be the way forward and it seems that it will replace most of my old `include` macros. I will add my thoughts to the other issue.
Shall we close this issue?

**Comment by mojavelinux** [comment]
Once something like a subdocument macro is available (which could also be done currently via an extension), I'm confident it will become more popular than the include directive. The include directive has its purpose (such as including CSV data, shared attribute entries, or output text), but that purpose is distinct and has inherent limitations (as I have described above).

I don't think we need to close this issue yet as it still depends on how the other issue evolves. But I do want to direct comments to #3603.

**Comment by SeidChr** [comment]
I had a connected problem in #4614
What i wanted was the preview for asciidoc documents to be working in VSCODE (asciidoctorjs) while also properly converting the whole document structure using asciidoctor.
The main problem here was the imagesdir, which was a different relative path depending on the file you are previewing. Thus images were not found.
For asciidoctor there was an attribute i could just pass on the commandline. For asciidoctorjs i found a way to pass it in the settings:

```json
{
  "asciidoc.extensions.enableKroki": true,
  "asciidoc.preview.asciidoctorAttributes": {
      "kroki-server-url": "https://################",
      "documentroot": "${workspaceFolder}/asciidoc/root/src",
      "imagesdir": "${workspaceFolder}/asciidoc/root/src/images",
  },
}
```

now the second problem i still have, is the above seen document root. Unfortunately the VSCODE ${workspaceFolder} inserts a / at the beginning of the string (which leads to `/c:/...` on windows), which causes the document lookup to fail. (not the images lookup though?!)

having this would allow me to just reference a section header like this, in every subsection document, without any adaptation, and thus have minimal effort when moving sections arround or creating new ones:
```
include::{documentroot}/_section.adoc[]
```

passing the documentroot to asciidoctor in the build process is easy enough.
Sorry, these problems are not really related to asciidoctor. I just needed a place to communicate my pain ;) Maybe somebody has an idea?

**Comment by mojavelinux** [comment]
The correct place to communicate experiences and pains with using Asciidoctor is the project chat at https://chat.asciidoctor.org. The issue tracker is not a forum or a support portal. Please redirect your inquiry to the chat.
